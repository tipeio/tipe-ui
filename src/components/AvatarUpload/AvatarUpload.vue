<template>
  <!--UPLOAD-->
  <div v-if="!avatar" :data-tipe-ui="$options.name" class="add-avatar">
    <input type="file" name="uploadFieldName" :disabled="isSaving" @change="uploadAvatar"
      accept="image/*" class="input-file">  
    <type-icon
      icon="avatar-upload"
      :width="styleObject.width"
      :height="styleObject.height"
    />
    <p v-if="isInitial" class="subtitle">Add avatar</p>
    <p v-else-if="isSaving" class="subtitle"> Uploading avatar...</p>
    <p v-else-if="uploadError" class="error">
      Accpeted formats: JPG, JPEG, GIF or PNG
    </p>
  </div>
  <!--success-->
  <img v-else :src="avatar" />
</template>

<script>
import TypeIcon from '../Icon'
import vueTypes from 'vue-types'
import { isImage } from '../../utils'

export default {
  name: 'AvatarUpload',
  components: {
    TypeIcon
  },
  props: {
    avatar: vueTypes.string,
    size: vueTypes.string.def('150')
  },
  data() {
    return {
      uploadError: null,
      isSaving: null,
      isInitial: !this.avatar
    }
  },
  computed: {
    styleObject() {
      return {
        width: this.size + 'px',
        height: this.size + 'px'
      } 
    }
  },
  methods: {
    uploadAvatar(e) {
      this.isInitial = false
      this.isSaving = true

      const file = e.target.files[0]
      if (!file.type || !isImage(file.type)) {
        this.isSaving = false
        this.uploadError = true
        return
      }

      this.$emit('onUpload', file)
    }
  }
}
</script>

<style scoped lang="postcss">
.input-file {
  opacity: 0; /* invisible but it's there! */
  width: 100%;
  height: 200px;
  position: absolute;
  cursor: pointer;
}
.add-avatar {
  display: grid;
  cursor: pointer;
  position: relative;
  width: 100%;
  grid-row-gap: 1rem;
  justify-content: center;
  align-content: start;

  & .circle {
    height: 100px;
    width: 100px;
    border-radius: 100%;
    background-color: var(--gray-blue-light);
  }

  & .outside {
    height: 100px;
    width: 100px;
    border-radius: 100%;
    border: var(--text-gray-light) dashed 1px;
  }

  & .subtitle {
    font-size: 1rem;
    color: var(--text-gray-normal);
    margin: 0;
    text-align: center;
  }

  & .error {
    font-size: 1rem;
    color: var(--error);
    margin: 0;
    text-align: center;
  }
}
</style>
