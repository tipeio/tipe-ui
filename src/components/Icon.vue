<template>
  <div
    :class="size"
    :data-tipe-ui="$options.name"
  />
</template>

<script>
import vueTypes from 'vue-types'

const sizes = ['small', 'medium', 'large']

export default {
  name: 'TipeIcon',
  props: {
    size: vueTypes.oneOf(sizes).def('medium')
  }
}
</script>

<style scoped>
[data-tipe-ui='TipeIcon'] {
  background-color: red;
}

[data-tipe-ui='TipeIcon'].small {
  height: 1rem;
  width: 1rem;
}

[data-tipe-ui='TipeIcon'].medium {
  height: 2rem;
  width: 2rem;
}

[data-tipe-ui='TipeIcon'].large {
  height: 3rem;
  width: 3rem;
}
</style>
