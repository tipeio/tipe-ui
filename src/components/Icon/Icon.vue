<template>
  <object
    :data="src"
    type="image/svg+xml">
    <img :src="src" >
  </object>
</template>

<script>
import vueTypes from 'vue-types'

export default {
  name: 'TipeIcon',
  props: {
    icon: vueTypes.string
  },
  computed: {
    src() {
      return require(`../../icons/${this.icon}.svg`)
    }
  }
}
</script>
