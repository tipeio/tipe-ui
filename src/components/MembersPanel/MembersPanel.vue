<template>
  <layout-card>
    <div class="grid">
      <div class="flex-row">
        <div class="flex-column">
          <p class="text">Members</p>
          <p class="small-text">Description text goes here</p>
        </div>
        <tipe-button icon-before="plus">Add Member</tipe-button>
      </div>
      <tipe-table :members="members" />
    </div>
  </layout-card>
</template>

<script>
import { TipeTable } from '.'
import { createManyMocks, user } from '../../mocks'
import LayoutCard from '../LayoutCard'
import TipeButton from '../Button'

export default {
  name: 'MembersPanel',
  components: { LayoutCard, TipeTable, TipeButton },
  data() {
    return {
      members: createManyMocks(user, 12)
    }
  }
}
</script>


<style lang='postcss' scoped>
.grid {
  display: grid;
  grid-template-rows: 4rem auto;
  width: 100%;
}

.flex-row {
  display: flex;
  max-width: 100%;
  justify-content: space-between;
  align-items: center;
  padding: 1.125rem;

  & .text {
    color: var(--darkPurple);
    font-size: 0.875rem;
    margin: 0;
    margin-bottom: 3px;
  }

  & .small-text {
    color: var(--midPurple);
    font-size: 0.75rem;
    margin: 0;
  }
}
.flex-column {
  display: flex;
  flex-direction: column;
}
</style>
