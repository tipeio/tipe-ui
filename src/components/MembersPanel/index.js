export TableRow from './TableRow'
export TipeTable from './Table'
export { default } from './MembersPanel'
