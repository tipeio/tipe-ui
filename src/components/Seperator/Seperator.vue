<template>
  <div :data-tipe-ui="$options.name" />
</template>

<script>
export default {
  name: 'TipeSeperator'
}
</script>

<style scoped>
[data-tipe-ui='TipeSeperator'] {
  height: 1px;
  background-color: var(--gray-light);
}
</style>
