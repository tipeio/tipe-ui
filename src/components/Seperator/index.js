export { default } from './Seperator.vue'
