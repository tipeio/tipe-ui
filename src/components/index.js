export const components = {
  Accordian: require('./Accordian'),
  AvatarUpload: require('./AvatarUpload'),
  Breadcrumbs: require('./Breadcrumbs'),
  Button: require('./Button'),
  Card: require('./Card'),
  Checkbox: require('./Checkbox'),
  Dialog: require('./Dialog'),
  DocumenEditor: require('./DocumentEditor'),
  DropDown: require('./DropDown'),
  Field: require('./Field'),
  Finder: require('./Finder'),
  Icon: require('./Icon'),
  IconBadge: require('./IconBadge'),
  Image: require('./Image'),
  Markdown: require('./Markdown'),
  Message: require('./Modal'),
  NumberInput: require('./NumberInput'),
  ProfileImage: require('./ProfileImage'),
  Progress: require('./Progress'),
  Radio: require('./Radio'),
  Scrolltable: require('./Scrollable'),
  Search: require('./Search'),
  Select: require('./Select'),
  SettingsMenu: require('./SettingsMenu'),
  Sidebar: require('./Sidebar'),
  Switch: require('./Switch'),
  Tabs: require('./Tabs'),
  TabsPanel: require('./TabsPanel'),
  TextArea: require('./Textarea'),
  TextInput: require('./TextInput')
}
