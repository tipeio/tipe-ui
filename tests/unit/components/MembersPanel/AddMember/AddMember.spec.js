import { shallowMount } from '@vue/test-utils'
import { AddMember } from '@/components/MembersPanel'
import { createRenderer } from 'vue-server-renderer'

import Icon from '@/components/Icon'
import TipeButton from '@/components/Button'
import Seperator from '@/components/Seperator'
import AvatarUpload from '@/components/AvatarUpload'
import TipeField from '@/components/Field'
import TipeSelect from '@/components/Select'

describe('AddMember.vue', () => {
  it('renders', () => {
    const renderer = createRenderer()
    const wrapper = shallowMount(AddMember)
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })
  it('has correct data-tipe-ui attibute', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.attributes()['data-tipe-ui']).toBe('AddMember')
  })
  it('should render the Icon component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(Icon)).toBe(true)
  })
  it('should render the TipeButton component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(TipeButton)).toBe(true)
  })
  it('should render the Seperator component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(Seperator)).toBe(true)
  })
  it('should render the AvatarUpload component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(AvatarUpload)).toBe(true)
  })
  it('should render the TipeField component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(TipeField)).toBe(true)
  })
  it('should render the TipeSelect component', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.contains(TipeSelect)).toBe(true)
  })
  it('should have the grid class', () => {
    const wrapper = shallowMount(AddMember)
    expect(wrapper.classes()).toContain('grid')
  })
  it('should invalidate the email input', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({ emailValue: '454' })
    wrapper.vm.emailValidate()
    expect(wrapper.vm.emailStatus).toBe('error')
  })
  it('should validate the email input', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({ emailValue: 'olivia@tipe.io' })
    wrapper.vm.emailValidate()
    expect(wrapper.vm.emailStatus).toBe('success')
  })
  it('should invalidate the name input', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({ nameValue: '' })
    wrapper.vm.nameValidate()
    expect(wrapper.vm.nameStatus).toBe('error')
  })
  it('should validate the name input', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({ nameValue: 'Olivia Oddo' })
    wrapper.vm.nameValidate()
    expect(wrapper.vm.nameStatus).toBe('success')
  })
  it('should not be able to submit the form', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailStatus: '',
      roleValue: { label: 'Member', value: 'member' },
      nameStatus: ''
    })
    expect(wrapper.vm.canSubmit).toBe(false)
  })
  it('should not be able to submit the form with invalid email', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailStatus: 'error',
      roleValue: { label: 'Member', value: 'member' },
      nameStatus: 'success'
    })
    expect(wrapper.vm.canSubmit).toBe(false)
  })
  it('should not be able to submit the form with invalid name', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailStatus: 'success',
      roleValue: { label: 'Member', value: 'member' },
      nameStatus: 'error'
    })
    expect(wrapper.vm.canSubmit).toBe(false)
  })
  it('should be able to submit the form', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailStatus: 'success',
      roleValue: { label: 'Member', value: 'member' },
      nameStatus: 'success'
    })
    expect(wrapper.vm.canSubmit).toBe(true)
  })
  it('should reset the form', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.setData({
      emailValue: 'olivia.oddo@gmail.com',
      roleValue: { label: 'Member', value: 'member' },
      nameValue: 'Olivia'
    })
    wrapper.vm.resetForm()
    expect(wrapper.vm.roleValue).toEqual({ label: 'Owner', value: 'owner' })
    expect(wrapper.vm.nameValue).toBe('')
    expect(wrapper.vm.emailValue).toBe('')
  })
  it('should update the name value', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.vm.onChangeName('Olivia')
    expect(wrapper.vm.nameValue).toEqual('Olivia')
  })
  it('should update the email value', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.vm.onChangeEmail('olivia.oddo@gmail.com')
    expect(wrapper.vm.emailValue).toEqual('olivia.oddo@gmail.com')
  })
  it('should update the role value', () => {
    const wrapper = shallowMount(AddMember)
    wrapper.vm.onChangeRole({ label: 'Owner', value: 'owner' })
    expect(wrapper.vm.roleValue).toEqual({ label: 'Owner', value: 'owner' })
  })
})
