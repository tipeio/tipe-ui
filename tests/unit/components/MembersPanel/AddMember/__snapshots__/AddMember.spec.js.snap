// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMember.vue renders 1`] = `
<div data-tipe-ui="AddMember" class="grid">
  <div class="close">
    <TipeIcon-stub icon="close" height="1rem" width="1rem" color="#000" __iconLoader__="function _default(icon) {
  return require(&quot;../../icons/&quot;.concat(icon, &quot;.svg&quot;)).default;
}"></TipeIcon-stub>
  </div>
  <div class="content">
    <p class="title">Add Member</p>
    <AvatarUpload-stub></AvatarUpload-stub>
    <form>
      <TipeField-stub name="name" field="[object Object]" label="Name" status="" errorMessage="" successMessage="" warningMessage="" class="name">
        <TipeTextInput-stub label="" placeholder="" name="" size="medium" tabindex="0"></TipeTextInput-stub>
      </TipeField-stub>
      <TipeField-stub name="email" field="[object Object]" label="Email" status="" errorMessage="" successMessage="" warningMessage="" class="email">
        <TipeTextInput-stub label="" placeholder="" name="" size="medium" tabindex="0"></TipeTextInput-stub>
      </TipeField-stub>
      <TipeField-stub name="" field="[object Object]" label="Role" status="" errorMessage="" successMessage="" warningMessage="" class="role">
        <TipeSelect-stub type="" placeholder="" name="" size="medium" tabindex="0"></TipeSelect-stub>
      </TipeField-stub>
      <div class="seperator">
        <TipeSeperator-stub></TipeSeperator-stub>
      </div>
      <div class="add-button">
        <TipeButton-stub color="primary" iconBefore="" iconAfter="" name="" value="Submit" disabled="disabled" size="large" tabindex="0" type="submit">ADD MEMBER</TipeButton-stub>
      </div>
    </form>
  </div>
</div>
`;
